### PR TITLE
Better documentation on what is good and bad

### PR DESCRIPTION
**Pull request by yosifkit**
This is a start of defining what we look for in PRs to the Official Images.  Hopefully we can make it easier for everyone involved.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -6,16 +6,77 @@
...
 
-Be sure to familiarize yourself with the [Guidelines for Creating and Documenting Official Repositories](https://docs.docker.com/docker-hub/official_repos/) and the [Best practices for writing Dockerfiles](https://docs.docker.com/articles/dockerfile_best-practices/) in the Docker documentation. These will be the foundation of the review process performed by the official images maintainers. If you'd like the review process to go more smoothly, please ensure that your `Dockerfile`s adhere to all the points mentioned there before submitting a pull request.
+Be sure to familiarize yourself with [Official Repositories on Docker Hub](https://docs.docker.com/docker-hub/official_repos/) and the [Best practices for writing Dockerfiles](https://docs.docker.com/articles/dockerfile_best-practices/) in the Docker documentation. These will be the foundation of the review process performed by the official images maintainers. If you'd like the review process to go more smoothly, please ensure that your `Dockerfile`s adhere to all the points mentioned there, as well as [below](README.md#quality-assurance), before submitting a pull request.
```
Comment (tianon) [inline README.md]: Whoops, `README.md#quality-assurance` should just be `#review-guidelines`.

File: README.md [hunk 2]
```diff
@@ -6,16 +6,77 @@
...
+
+#### Repeatability
+
+Rebuilding the same `Dockerfile` should result in the same version of the image being packaged, even if the second build happens several versions later, or the build should fail outright, such that an inadvertent rebuild of a `Dockerfile` tagged as `0.1.0` doesn't end up containing `0.2.3`. For example, if using `apt` to install the main program for the image, be sure to pin it to a specific version (ex: `... apt-get install -y my-package=0.1.0 ...`). For dependent packages installed by `apt` there is not usually a need to pin them to a version.
+
+#### Consistency
+
+All official images should provide a consistent interface. A beginning user should be able to `docker run official-image bash` without needing to learn about `--entrypoint`. A minimal image that only contains the main executable and its linked libraries and thus no shell would be exempt. It is also nice for advanced users to take advantage of entrypoint, so that they can `docker run official-image --arg1 --arg2` without having to specify the binary to execute.
```
Comment (psftw) [inline README.md]: I think I know what you are trying to say, but I could see people hitting this section and getting confused.  Maybe list the three acceptable solutions and when they should be used? IF I have it correct: 1. No entrypoint 2. Entrypoint for initialization, 3. Entrypoint for minimal image that only takes parameters

File: README.md [hunk 3]
```diff
@@ -6,16 +6,77 @@
...
+
+#### Clarity
+
+Try to make the `Dockerfile` easy to understand/read. It may be tempting, for the sake of brevity, to put complicated initialization details into a standalone script and merely add a `RUN` command in the `Dockerfile`. However, this causes the resulting `Dockerfile` to be overly opaque, and such `Dockerfile`s are unlikely to pass review. Instead, it it recommended to put all the commands for initialization into the `Dockerfile` as appropriate `RUN` or `ENV` command combinations. To find good examples, look at the current official images.
```
Comment (psftw) [inline README.md]: What about all the entrypoint scripts that do initialization and other complex stuff.  There is a subtle difference worth calling out here for that case maybe?

File: README.md [hunk 4]
```diff
@@ -6,16 +6,77 @@
...
+
+#### Security
+
+Version bumps and security fixes need to be attended to in a timely manner. Additionally, the `Dockerfile` should be designed to prevent man-in-the-middle attacks: using https where possible; importing PGP keys with the full fingerprint in the Dockerfile to check package signing; embedding checksums directly in the `Dockerfile` if PGP signing is not provided. When importing PGP keys, we recommend using the [high-availability server pool](https://sks-keyservers.net/overview-of-pools.php#pool_ha) from sks-keyservers (`ha.pool.sks-keyservers.net`). Here are a few good and bad examples:
```
Comment (psftw) [inline README.md]: The first sentence feels a little bit out of place. So far in the "Review Guidelines" section we've talked about additional quality requirements for Dockerfiles, but then this lone sentence tells me that updates need to be done in a timely manner.  Maybe talk about update frequency in the Repeatability section?  The rest of the Security section is really great :+1:

File: README.md [hunk 5]
```diff
@@ -6,16 +6,77 @@
...
+		# install
+	```
+
+-	**Best**: *full key fingerprint imported to apt-key which will check signatures when packages are downloaded and installed.*
+	``` Dockerfile
+	RUN apt-key adv --keyserver ha.pool.sks-keyservers.net --recv-keys 492EAFE8CD016A07919F1D2B9ECBEC467F0CEB10
+	RUN echo "deb http://repo.mongodb.org/apt/debian wheezy/mongodb-org/$MONGO_MAJOR main" > /etc/apt/sources.list.d/mongodb-org.list
+	RUN apt-get update \
+		&& apt-get install -y mongodb-org=$MONGO_VERSION \
+		&& rm -rf /var/lib/apt/lists/* \
```
Comment (psftw) [inline README.md]: What's this `rm -rf /var/lib/apt/lists/*` business and why should I use it??? :trollface:

File: README.md [hunk 6]
```diff
@@ -6,16 +6,77 @@
...
+	``` Dockerfile
+	# gpg: key F73C700D: public key "Larry Hastings <larry@hastings.org>" imported
+	RUN gpg --keyserver ha.pool.sks-keyservers.net --recv-keys 97FC712E4C024BBEA48A61ED3A5CA953F73C700D
+	RUN curl -fSL "https://www.python.org/ftp/python/$PYTHON_VERSION/Python-$PYTHON_VERSION.tar.xz" -o python.tar.xz \
+		&& curl -fSL "https://www.python.org/ftp/python/$PYTHON_VERSION/Python-$PYTHON_VERSION.tar.xz.asc" -o python.tar.xz.asc \
+		&& gpg --verify python.tar.xz.asc \
+		# install
+	```
+
+#### Miscellaneous
```
Comment (psftw) [inline README.md]: Number 3 is meaningless as-is, and the first two points under Misc. could have their own sections above instead (?).

**Comment by psftw** [comment]
sorry, apparently I haven't figured out how to comment on specific lines/sections correctly.  My comments are mostly specific to the last line in each excerpt

**Comment by tianon** [comment]
@psftw want to re-review the exploded version, or would the squashed version be simpler?

**Comment by psftw** [comment]
@yosifkit @tianon LGTM.  I don't care how the commits look, but the current state of the branch is a big improvement. Thanks!

File: README.md [hunk 7]
```diff
@@ -6,24 +6,152 @@
...
+### Review Guidelines
 
-1.	issues with build repeatability (rebuilding the same `Dockerfile` resulting in the same version of the image being packaged, even if the second build happens several versions later, such that an inadvertent rebuild of a `Dockerfile` tagged as `0.1.0` doesn't end up containing `0.2.3`, for example)
-2.	things that cause technical issues based on our experience and familiarity (unnecessary `COPY` falls in here)
-3.	things that cause maintenance issues (like hard-coding version numbers in more than one place instead of using `ENV`, for example, which inevitably leads to overlooking necessary changes during an image update)
-4.	things that cause usability or consistency issues
+Because the official images are intended to be learning tools for those new to Docker as well as the base images for advanced users to build their production releases, we review each proposed `Dockerfile` to ensure that it meets a minimum standard for quality and maintainability. While some of that standard is hard to define (due to subjectivity), as much as possible is defined here, while also adhering to the "Best Practices" where appropriate.
```
Comment (md5) [inline README.md]: I'm not sure "due to subjectivity" is the best thing to say here. I'd say something like "due to the fast-evolving nature of the Docker ecosystem", but that's too long.

Comment (md5) [inline README.md]: "as much as possible is defined here" is confusing. Perhaps "as much of that standard as possible is defined here".

> While some of that standard is subjective due to the fast-evolving nature of the Docker ecosystem, as much of the standard as is practical to put in writing is defined here, while also adhering to the "Best Practices" where appropriate.

Comment (md5) [inline README.md]: That still sucks...

**Comment by md5** [comment]
LGTM

Just ignore my line comments. Further PRs can be used to refine.

**Comment by tianon** [comment]
Good enough for me. :+1:

Looking forward to follow up PRs. :smile:

**Pull request merged**